lib/fluent/plugin/out_http.rb

```diff
@@ -325,6 +330,12 @@ def create_request(chunk, uri)
       set_headers(req, uri, chunk)
       req.body = @json_array ? "[#{chunk.read.chop}]" : chunk.read
 
+      if @compress == :gzip
+        gz = Zlib::GzipWriter.new(StringIO.new)
+        gz << req.body
+        req.body = gz.close.string
+      end
+
```

Comment:
We need to consider the possibility that the chunk is already compressed ([compress option of buffer](https://docs.fluentd.org/configuration/buffer-section#buffering-parameters)).

Example: `out_file` changes the behavior depending on `@buffer.compress`.

https://github.com/fluent/fluentd/blob/c0cd1e6960290954b442ea24077a80cee9229d41/lib/fluent/plugin/out_file.rb#L214-L222

Example: `out_forward` forces `@buffer.compress` to `gzip` when own `compress` is `gzip`.

https://github.com/fluent/fluentd/blob/c0cd1e6960290954b442ea24077a80cee9229d41/lib/fluent/plugin/out_forward.rb#L254-L258

https://github.com/fluent/fluentd/blob/c0cd1e6960290954b442ea24077a80cee9229d41/lib/fluent/plugin/out_forward.rb#L664-L668

Like `out_forward`, we could force `@buffer.compress` to `gzip` when `out_http`'s `compress` is `gzip`.
(Then, it would be enough to read chunk as is like `chunk.read(compressed: @compress)`.)
On the other hand, we should note that the compression efficiency decreases when data is compressed frequently. (`out_file` has [recompress option](https://docs.fluentd.org/output/file#recompress) for this.)

Anyway, we need to take into account the compression of buffer to some extent.

Reply:
The current PR's specification is to always recompress data regardless of the Buffer's compression settings, right?
(`chunk.read` without arguments will decompress data if the chunk is compressed).

If this is the intended specification, the current implementation would be fine.

Reply:
At current yes it will decompress the buffer, format the response, and recompress before sending.

I think trying to send already compressed chunks straight from the buffer might interfere with the json_array setting, or at least theres a case to be handled there.

I'll take a look and see what improvements can be made here

Reply:
Thanks!
Note: https://github.com/fluent/fluentd/pull/4528#discussion_r1639078250